### PR TITLE
Adding migration generator

### PR DESCRIPTION
**Pull request by thecatwasnot**
No options currently used

`bundle exec lotus generate migration migration_name` creates `db/migrations/<timestamp>_migration_name.rb`

`bundle exec lotus generate migration slice_name migration_name` creates `apps/slice_name/db/migrations/<timestamp>_migration_name.rb`

### CODE REVIEW

File: lib/lotus/generators/migration.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      def start
+        opts = {
+          #app_name:         app_name,
+          migration_name:   migration_name,
+          migration_time:   migration_time,
+          migration_class:  migration_class,
+        }
+
+        templates = {
+          'migration.rb.tt' => "db/migrate/#{ opts[:migration_time] }_#{ opts[:migration_name] }.rb"
```
Comment (joneslee85) [inline lib/lotus/generators/migration.rb]: `#{ opts[:migration_time] }_#{ opts[:migration_name] }` can be extracted to a private method

File: lib/lotus/commands/generate/migration.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    module Generate
+      class Migration 
+        
+        attr_reader :app_name, :migration_name, :migration_time, :migration_class, :cli, :source, :target
+
+        def initialize(name, environment, cli)
+          name = Lotus::Utils::String.new(name)
+          #@app_name = ""
+          @migration_name = name.underscore
+          @migration_time = Time.now.to_i.to_s
```
Comment (joneslee85) [inline lib/lotus/commands/generate/migration.rb]: Timestamp should be human readable, what do you think about formatting it as `ddmmyyyy`?

File: test/commands/generate/migration_test.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  before do
+    create_temporary_dir
+  end
+
+  after do
+    chdir_to_root
+  end
+
+  describe 'migration' do
```
Comment (joneslee85) [inline test/commands/generate/migration_test.rb]: how about `describe '#start'`?

File: lib/lotus/commands/generate/migration.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        end
+
+        def start
+          @command.start
+        end
+
+        private
+        
+        def generate_timestamp
+          Time.now.strftime("%d%m%Y")
```
Comment (AlfonsoUceda) [inline lib/lotus/commands/generate/migration.rb]: It should be changed, better this for better sorting:

`Time.now.strftime("%Y%m%d")` #=> 20150117

File: lib/lotus/cli/generate.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+require 'thor'
+require 'lotus/environment'
+
+
+module Lotus
+  class Generate < Thor
+    include Thor::Actions
+    desc 'migration NAME', 'generates a model migration'
```
Comment (AlfonsoUceda) [inline lib/lotus/cli/generate.rb]: space above please, between this line and include

**Comment by AlfonsoUceda** [comment]
:+1: perfect for me except above comments

File: lib/lotus/commands/generate/migration.rb (hunk 4 above)
Comment (thecatwasnot) [inline lib/lotus/commands/generate/migration.rb]: Actually it should probably be YmdHs to keep unique.

Comment (AlfonsoUceda) [inline lib/lotus/commands/generate/migration.rb]: perfect ;)

File: lib/lotus/commands/generate/migration.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+module Lotus
+  module Commands
+    module Generate
+      class Migration 
+        
+        attr_reader :app_name, :migration_name, :migration_time, :migration_class, :cli, :source, :target
+
+        def initialize(name, environment, cli)
+          name = Lotus::Utils::String.new(name)
+          #@app_name = ""
```
Comment (joneslee85) [inline lib/lotus/commands/generate/migration.rb]: this could be removed

File: lib/lotus/commands/generate/migration.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+module Lotus
+  module Commands
+    module Generate
+      class Migration 
+        
+        attr_reader :migration_name, :migration_time, :migration_class, :cli, :source, :target
+
+        def initialize(name, environment, cli)
+          name = Lotus::Utils::String.new(name)
+          @migration_name = name.underscore
```
Comment (joneslee85) [inline lib/lotus/commands/generate/migration.rb]: What do you think about `Tell, Don't Ask`? I think it'd better to have a private method `generate_name` here.

File: lib/lotus/generators/migration.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+        templates.each do |src, dst|
+          cli.template(source.join(src), target.join(dst), opts)
+        end
+      end
+
+      private
+
+      def migration_file_name(time, name)
+        "#{time}_#{name}"
```
Comment (joneslee85) [inline lib/lotus/generators/migration.rb]: Why don't we access directly to `migration_time` and `migration_name` here?

File: test/commands/generate/migration_test.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+  before do
+    create_temporary_dir
+  end
+
+  after do
+    chdir_to_root
+  end
+
+  describe 'start' do
```
Comment (joneslee85) [inline test/commands/generate/migration_test.rb]: How about following convention `#` prefix to denote a method? That is `#start` instead of `start`

**Comment by joneslee85** [comment]
except above comments and failed build, this is :+1: 

Pending discussion: In order to generate migration for `lib/chirp` and `apps/web`, we need a reliable way to detect the app name. Suggested solution is having app name defined in `.lotusrc` file. I am not really a fan of this approach, I prefer an alternative and implicit way. 

Here is my proposal:

* Enforce naming convention for `lib/chirp`, that is renaming `lib/chirp.rb` to `lib/application.rb`
* Introduce options to set migration path for `Lotus::Model` and `Lotus::Application`


@jodosha thoughts?

**Comment by thecatwasnot** [comment]
My current feeling is that the application db migratioins should live in `db/migrate` and slice specific migrations should live in `apps/web/db/migrate`
IMHO that would be the least surprising option.
I can add the slice migration generator with `bundle exec lotus generate migration web migration_name`

**Comment by joneslee85** [comment]
@thecatwasnot `db/migrate` seems to be poor choice of naming, as suggested, I want to rename it to `db/migrations`, can you do some magic and make it happen? TIA

**Comment by AlfonsoUceda** [comment]
I think we should cover the 2 scenarios:
`bundle exec lotus generate migration migration_name` and `bundle exec lotus generate migration slice_name migration_name`

**Comment by thecatwasnot** [comment]
this is in progress guys, should be done tonight

**Comment by thecatwasnot** [comment]
#256 has more features & is up to date

**Pull request closed without merging**